### PR TITLE
/ and other ISSUE! chars, make # space (not empty)

### PR DESCRIPTION
**Pull request by hostilefork**
Rebol2 and R3-Alpha break issues to become refinements at slash:

    rebol2>> load "#ab/cd"
    == [#ab /cd
    ]

This changes it to follow the same load rules as FILE!, and include
the slashes:

    >> x: load "#ab/cd"
    == #ab/cd

    >> type of x
    == #[datatype! issue!]

It also makes more characters in ISSUE! legal.  This is designed
to allow it evolve into the "cleaner" form of CHAR!, ultimately
replacing the CHAR! datatype entirely.

    >> ensure issue! #:
    == #:

    >> ensure issue! #<
    == #<

While it looks sensible that `#;` would be an ISSUE!, the ; is
a comment anywhere at this time.  In order to help catch cases
that would forget and use `#;` like `#:` and cause hard to find
bugs of code getting commented out, this makes that case error.

It also makes the choice that `#` means an issue containing a space,
not the empty issue.  This offers a convenient alternative to
expressing space, that is shorter than #" " or the word `space`.

### CODE REVIEW

**Comment by hostilefork** [comment]
So @rgchris, I don't know where you stand on working with R3C (**it would help if I knew!**).  But I tried to factor this out to make it easy for you to apply, if you chose to.  I think by and large things are falling into place in a way consistent with what you have historically advocated for... e.g. FILE! and ISSUE! having internal slashes.

This is all related closely to the [ISSUE! and CHAR! unification into one atomic UTF-8 optimized immutable type](https://forum.rebol.info/t/removing-issue-from-the-any-word-category/1106/3).

I want to be really certain about the # decaying to just space part.  It gives an alternative to BLANK! as space:

    >> unspaced ["controversial" _ "behavior"]
    convtroversial behavior  ; !!! we've wondered if this was right

    >> unspaced ["replacement" # "behavior"]
    replacement behavior  ; this may be a more solid answer

It doesn't vanish quite as much as the BLANK! does for spacing.  But there's some benefit in that, as it's kind of a blocky "negative space" space.  But the fact that # would really be the space character in contexts it was found that were character related would make it less of a surprise:

    >> second "a b"
    == #

I'd whimsically toyed with the idea of contexts something like that returning BLANK! before, e.g. TO BLOCK! of TEXT!

     >> to block! "ab cd"
     == [#"a" #"b" _ #"c" #"d"]

It made some ideas better, but that's no match for:

     >> to block! "ab cd"
     == [#a #b # #c #d]

But making all dialects equate BLANK! with how NULL is handled doesn't feel right, as I really want this to work:

       >> did parse [1 "a" 2] [integer! _ integer!]
       == #[true]

This makes PARSE suitable for "deconstructing" things, and underscore is the traditionally-used "I don't care" of deconstruction.

Anyway, food for thought and hopefully adding up changes you will approve of.

**Comment by hostilefork** [comment]
> It made some ideas better, but that's no match for:
>
>      >> to block! "ab cd"
>      == [#a #b # #c #d]

There's definitely a lot of improvement there cutting down on the quotes.  But as I push this through a bit further, I'm finding myself kind of less happy with the # - as - space concept.  There's something off balance in the 1-charness of it.

Something I'm wondering is if caret standing alone could start one of these ISSUECHAR!s when escaping was involved:

      >> to block! "ab cd"
      == [#a #b ^_ #c #d]

That feels more "balanced", to have things like **`^/`** or **`^(1C)`** standing on its own as a newline, lexically.

So...while I have conflicted feelings about BLANK!-as-space vs. BLANK!-as-opt-out in the DELIMIT "dialect", the scenario I give with PARSE acting as SKIP speaks to the idea that I don't think blank needs to be a null equivalent...and something is lost representationally when it is.

But if ^_ is space, then that lines up nicely with _ acting as a dialected space, as it was.

In other words: I'm very much liking this direction for the ISSUECHAR! (name pending), but I think in the context of seeing it "among its peers", # isn't suitable for a space.

But an empty issue char could be saved for representing 0, and have the positive benefit of being something that is visibly unique and yet does not freely append itself to strings.  See the use here of "sending into space" (but it would no longer be space):

https://forum.rebol.info/t/sending-values-into-space/1347

**Comment by iArnold** [comment]
> It also makes the choice that # means an issue containing a space,
> not the empty issue. This offers a convenient alternative to
> expressing space, that is shorter than #" " or the word space.

And #{} would be the empty issue like <{}> the empty TAG?

**Comment by hostilefork** [comment]
> And #{} would be the empty issue like <{}> the empty TAG?

I'm leaning that <{}> as the empty tag is probably the way to go.  But I've mentioned that trying out # as space in practice wasn't as pleasing as I at first thought.

So # being an empty "issuechar!" does have applications, like what I'm currently calling "blackhole" usage:

https://forum.rebol.info/t/sending-values-into-space/1347

But it could also be a synonym for #"^(00)", or #^(00) or ^(00) ... e.g. an issuechar! that converts to codepoint 0.  One benefit of that is that it becomes "toxic" in the process, so that you can't append it to strings.  Which makes it a better outlier type for such purposes.

**Comment by hostilefork** [comment]
Treating # as the empty issue.  Perhaps it could ultimately double as the "codepoint 0" CHAR!, e.g. **append &{AABB} #** would produce &{AABB00} under the new rules.  This would avoid creating a stringlike class that actually materially contained 0 bytes, which I'm trying to avoid.

Committed here:

https://github.com/metaeducation/ren-c/commit/4f3c86f256000cc55992d780d2dbd488962250b8

So to sum up the good news: more characters open up for ISSUE! (like / and .) as they are no longer wanted for putting ISSUE! in PATH! or TUPLE!.  The "bad" news is that it will become an immutable class with optimized storage and no indexed position, so it will need to be converted to text for manipulations of that kind.

**Comment by giuliolunati** [comment]
Not sure about # being the "space issue"... it seems anti-intuitive to me...

**Comment by hostilefork** [comment]
> Not sure about # being the "space issue"... it seems anti-intuitive to me...

As I mentioned, that was abandoned.  The usage of BLANK! as a space in things like DELIMIT is now standardized, with it only applying to "source level blanks":

https://forum.rebol.info/t/treat-blank-s-from-variables-or-evaluation-like-null/1348
